File: LeetCode/P736Hard_Parse_Lisp_Expression/main.py
# Problem: https://leetcode.com/problems/parse-lisp-expression/description/

class Number:
    def __init__(self, val):
        self.val = val

    def evaluate(self, env):
        return self.val


class Var:
    def __init__(self, name):
        self.name = name

    def evaluate(self, env):
        return env[self.name]


class Add:
    def __init__(self, expr1, expr2):
        self.expr1 = expr1
        self.expr2 = expr2

    def evaluate(self, env):
        return self.expr1.evaluate(env) + self.expr2.evaluate(env)


class Mult:
    def __init__(self, expr1, expr2):
        self.expr1 = expr1
        self.expr2 = expr2

    def evaluate(self, env):
        return self.expr1.evaluate(env) * self.expr2.evaluate(env)


class Let:
    def __init__(self, bindings, expr):
        self.bindings = bindings
        self.expr = expr

    def evaluate(self, env):
        newenv = dict(env)
        for var, expr in self.bindings:
            newenv[var] = expr.evaluate(newenv)
        return self.expr.evaluate(newenv)


class Solution:
    def make_toks(self, expression, i=0):
        res = []
        # assume i points to open brace
        i += 1

        while i < len(expression):
            if expression[i] == ' ':
                i += 1
            elif expression[i] == '(':
                newres, i = self.make_toks(expression, i)
                res.append(newres)
            elif expression[i] == ')':
                return res, i+1
            else:
                nxt = min(filter(lambda x: x >= 0,
                                 (expression.find(' ', i), expression.find(')', i))))
                res.append(expression[i:nxt])
                i = nxt
        return res, i

    def parse(self, toks):
        if type(toks) == list:
            return self.parse_op(toks)
        else:
            return self.parse_literal(toks)

    def parse_op(self, toks):
        if toks[0] == 'let':
            bindings = []
            for i in range(1, len(toks)-2, 2):
                bindings.append((toks[i], self.parse(toks[i+1])))
            return Let(bindings, self.parse(toks[-1]))
        elif toks[0] == 'mult':
            return Mult(self.parse(toks[1]), self.parse(toks[2]))
        elif toks[0] == 'add':
            return Add(self.parse(toks[1]), self.parse(toks[2]))

    def parse_literal(self, tok):
        if tok.isnumeric() or (tok[0] == '-' and tok[1:].isnumeric()):
            return Number(int(tok))
        else:
            return Var(tok)

    def evaluate(self, expression: str) -> int:
        toks, _ = self.make_toks(expression)
        expr = self.parse(toks)
        return expr.evaluate({})

File: LeetCode/P736Hard_Parse_Lisp_Expression/main.cpp
#include <iostream>
#include <map>
#include <vector>
// Problem: https://leetcode.com/problems/parse-lisp-expression/

using namespace std;

// define headers
class Expr;
std::ostream &operator<<(std::ostream &os, const Expr *expr);
typedef map<string, int> var_map;

class Expr {
public:
  virtual int eval(var_map &map) = 0;
  virtual std::ostream &print(std::ostream &os) const = 0;
  virtual ~Expr(){};
};

class Number : public Expr {
private:
  int val;

public:
  Number(int val) : val(val) {}
  ~Number(){};
  int eval(var_map &vars) override { return val; }
  std::ostream &print(std::ostream &os) const override {
    os << "Number(" << val << ")";
    return os;
  };
};

class Variable : public Expr {
private:
  string name;

public:
  Variable(string name) : name(name) {}
  ~Variable(){};
  int eval(var_map &vars) override { return vars[name]; }
  std::ostream &print(std::ostream &os) const override {
    os << "Variable(" << '"' << name << '"' << ")";
    return os;
  }
};

class Add : public Expr {
private:
  Expr *expr1;
  Expr *expr2;

public:
  Add(Expr *expr1, Expr *expr2) : expr1(expr1), expr2(expr2) {}
  ~Add() {
    delete expr1;
    delete expr2;
  };
  int eval(var_map &vars) override {
    return expr1->eval(vars) + expr2->eval(vars);
  }
  std::ostream &print(std::ostream &os) const override {
    os << "Add(" << expr1 << ", " << expr2 << ")";
    return os;
  }
};

class Mult : public Expr {
private:
  Expr *expr1;
  Expr *expr2;

public:
  Mult(Expr *expr1, Expr *expr2) : expr1(expr1), expr2(expr2) {}
  ~Mult() {
    delete expr1;
    delete expr2;
  };
  int eval(var_map &vars) override {
    return expr1->eval(vars) * expr2->eval(vars);
  }
  std::ostream &print(std::ostream &os) const override {
    os << "Mult(" << expr1 << ", " << expr2 << ")";
    return os;
  }
};

class Let : public Expr {
private:
  vector<pair<string, Expr *>> bindings;
  Expr *inner_expr;

public:
  Let(vector<pair<string, Expr *>> &bindings, Expr *inner_expr)
      : bindings(bindings), inner_expr(inner_expr){};
  ~Let() {
    for (auto p : bindings) {
      delete p.second;
    }
    delete inner_expr;
  };
  int eval(var_map &vars) override {
    var_map newvars = vars;
    for (auto p : bindings) {
      newvars[p.first] = p.second->eval(newvars);
    }
    return inner_expr->eval(newvars);
  }
  std::ostream &print(std::ostream &os) const override {
    os << "Let({";
    for (auto p : bindings) {
      os << "(\"" << p.first << "\", " << p.second << "), ";
    }
    os << "}, " << inner_expr << ")";
    return os;
  }
};

std::ostream &operator<<(std::ostream &os, const Expr *expr) {
  return expr->print(os);
}

class Solution {
private:
  vector<string> make_toks(string &expression) {
    vector<string> res;
    int prev = 0;
    int i = 0;

    while (i < expression.size()) {
      if (expression[i] == '(') {
        res.push_back("(");
        i++;
        prev++;
      } else if (expression[i] == ')') {
        if (prev != i)
          res.push_back(expression.substr(prev, i - prev));
        res.push_back(")");
        i++;
        prev = i;
      } else if (expression[i] == ' ') {
        if (prev != i)
          res.push_back(expression.substr(prev, i - prev));
        i++;
        prev = i;
      } else {
        i++;
      }
    }

    return res;
  }

  Expr *parse(vector<string> &toks) {
    int i = 0;
    return parse(toks, i);
  }

  Expr *parse(vector<string> &toks, int &i) {
    if (toks[i] == "(") {
      return parse_op(toks, i);
    } else {
      return parse_lit(toks, i);
    }
  }

  Expr *parse_lit(vector<string> &toks, int &i) {
    if (isalpha(toks[i][0])) {
      // variable name
      return new Variable(toks[i++]);
    } else {
      // number
      return new Number(stoi(toks[i++]));
    }
  }

  Expr *parse_op(vector<string> &toks, int &i) {
    i++; // skip '('
    if (toks[i] == "add") {
      i++; // consume opname
      Expr *e1 = parse(toks, i);
      Expr *e2 = parse(toks, i);
      i++; // skip ')'
      return new Add(e1, e2);
    } else if (toks[i] == "mult") {
      i++; // consume opname
      Expr *e1 = parse(toks, i);
      Expr *e2 = parse(toks, i);
      i++; // skip ')'
      return new Mult(e1, e2);
    } else {
      // toks[i] == "let"
      i++; // consume opname
      vector<pair<string, Expr *>> bindings;
      while (true) {
        if (toks[i] != "(" && isalpha(toks[i][0]) && toks[i + 1] != ")") {
          string varname = toks[i++]; // consume varname
          Expr *e1 = parse(toks, i);
          bindings.push_back({varname, e1});
        } else {
          Expr *inner = parse(toks, i);
          i++; // skip ')'
          return new Let(bindings, inner);
        }
      }
    }
  }

  int evaluate(Expr *e) {
    var_map vars;
    return e->eval(vars);
  }

public:
  int evaluate(string expression) {
    vector<string> toks = make_toks(expression);
    Expr *expr = parse(toks);
    int res = evaluate(expr);
    delete expr;
    return res;
  }
};
